src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+    const postcodeRegex = new RegExp('([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|' +
+      '(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))' +
+      '\\s?[0-9][A-Za-z]{2})', 'g');
+
+    const match = postcode.match(postcodeRegex);
+    return match?.length === 1 && match[0] === postcode;
+  }
+
+  private getAddressTypesForSelect(addressTypes: AddressType[], isPrimaryAddress: boolean): SelectItem[] {
+    const allAddressTypes = addressTypes.map((at: AddressType) => ({ value: at.id, text: at.name, selected: false }));
+
+    return isPrimaryAddress
+      ? allAddressTypes
+      : allAddressTypes.filter(at => at.text.toLowerCase() !== this.visitOrContactUsAddressType.toLowerCase());
+  }
+
+  private checkErrorResponseForPostcodeErrors(error: AxiosError, addresses: DisplayCourtAddresses):
+    { primaryInvalid: boolean; secondaryInvalid: boolean; errors: string[] } {
+
+    let primaryPostcodeInvalid = false;
+    let secondaryPostcodeInvalid = false;
+    const errors: string[] = [];
+
+    // We expect an array of invalid postcodes in the body of the response
+    if (Array.isArray(error.response.data)) {
+      const invalidPostcodes = error.response.data as string[];
+
+      invalidPostcodes.forEach(invalidPostcode => {
+        if (invalidPostcode === addresses.primary?.postcode) {
+          primaryPostcodeInvalid = true;
+          errors.push(this.primaryAddressPrefix + this.postcodeNotFoundError);
+        } else if (invalidPostcode === addresses.secondary?.postcode) {
+          secondaryPostcodeInvalid = true;
+          errors.push(this.secondaryAddressPrefix + this.postcodeNotFoundError);
+        }
+      });
+    }
+
+    return { primaryInvalid: primaryPostcodeInvalid, secondaryInvalid: secondaryPostcodeInvalid, errors: errors };
+  }
+
+  private convertToDisplayAddresses(addresses: CourtAddress[]): DisplayCourtAddresses {
+    const courtAddresses: DisplayCourtAddresses = { primary: null, secondary: null };
+    if (addresses.length > 0) {
+      courtAddresses.primary = this.convertApiAddressToCourtAddressType(addresses[0]);
+    }
+    if (addresses.length > 1) {
+      courtAddresses.secondary = this.convertApiAddressToCourtAddressType(addresses[1]);
+    }
+    return courtAddresses;
+  }
+
+  private convertToApiType(courtAddresses: DisplayCourtAddresses): CourtAddress[] {
+    const apiAddresses: CourtAddress[] = [];
+    if (courtAddresses.primary) {
+      apiAddresses.push(this.convertCourtAddressToApiAddressType(courtAddresses.primary));
+    }
+    if (courtAddresses.secondary && courtAddresses.secondary.type_id && courtAddresses.secondary.address_lines &&
+      courtAddresses.secondary.town && courtAddresses.secondary.postcode) {
+      apiAddresses.push(this.convertCourtAddressToApiAddressType(courtAddresses.secondary));
+    }
+    return apiAddresses;
+  }
+
+  private convertCourtAddressToApiAddressType(courtAddress: DisplayAddress): CourtAddress {
```

Comment:
The methods here and the one below it are essentially the same, with the exception of a different parameter and response

Not one to worry about too much, but I think in Java you can have two types using the same extendable class, for example one might be (off the top of my head): 

`public abstract class Bird extends Animal` 

`public abstract class Doggo extends Animal` 

Is there something similar we can do here? Not to worry if not

Reply:
I had initially thought about doing something like this with the types but the part of each address that differs is the main body of the address so it's difficult to create a base type that passes the 'is a' test.

I've also not used 'get' accessors in a single type because it needs to be easily serializable and deserializable.